csi/node_server.go

```diff
@@ -147,45 +147,15 @@ func (ns *NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 		return nil, status.Errorf(codes.Aborted, "no %v workload pods for volume %v to be mounted: %+v", corev1.PodPending, volumeID, podsStatus)
 	}
 
-	if volumeCapability.GetBlock() != nil {
-		devicePath := getStageBlockVolumePath(stagingTargetPath, volumeID)
-		_, err := os.Stat(devicePath)
-		if err != nil {
-			if !os.IsNotExist(err) {
-				return nil, status.Errorf(codes.Internal, errors.Wrapf(err, "failed to stat device %s", devicePath).Error())
-			}
-			// Fall back to the controller endpoint if the device path under the stagingTargetPath doesn't exist
-			log.Infof("Device path %s doesn't exist, falling back to controller endpoint %s", devicePath, volume.Controllers[0].Endpoint)
-			devicePath = volume.Controllers[0].Endpoint
-		}
-
-		if err := ns.nodePublishBlockVolume(volumeID, devicePath, targetPath, mounter); err != nil {
-			log.WithError(err).Errorf("Failed to publish BlockVolume %s", volumeID)
-			return nil, err
-		}
-
-		log.Infof("Published BlockVolume %s", volumeID)
-		return &csi.NodePublishVolumeResponse{}, nil
-	}
-
-	// we validate the staging path to make sure the global mount is still valid
-	unstageVolumeRequired := true
-	if volume.DataEngine == string(longhorn.DataEngineTypeV2) {
-		err = fmt.Errorf("always unstage v2 volume %v", volumeID)
-	} else {
-		isMnt, err := ensureMountPoint(stagingTargetPath, mounter)
-		if err == nil && isMnt {
-			unstageVolumeRequired = false
-		}
-	}
-	if unstageVolumeRequired {
+	// It may be necessary to restage the volume before we can publish it. For example, sometimes kubelet calls
+	// NodePublishVolume without calling NodeStageVolume. According to the CSI spec, we should be able to respond with
+	// FailedPrecondition and expect kubelet to call NodeStageVolume again, but as of Kubernetes v1.27 it does not.
+	isBlock := volumeCapability.GetBlock() != nil
+	restageRequired, err := restageRequired(volume, volumeID, stagingTargetPath, mounter, isBlock)
+	if restageRequired {
 		msg := fmt.Sprintf("Staging target path %v is no longer valid for volume %v", stagingTargetPath, volumeID)
 		log.WithError(err).Error(msg)
```

Comment:
It seems reasonable that this is not really an Error level log, but just Warning like the one immediately following it.  Is it possible to all log.WithError(err).Warning(msg) or failing that, to format the `err` explanation from restageRequired() into `msg`?

Reply:
Yeah, the ` log.WithError(err).Warning(msg)` seemed reasonable. Adopted it in an amend.